tools/eos-profile-tool/eos-profile-cmd-diff.c

```diff
@@ -0,0 +1,422 @@
+          gvdb_table_free (db);
+          return 1;
+        }
+
+      ForeachClosure clos = { 
+        .filename = opt_files[i],
+        .probes = probes,
+      };
+
+      eos_profile_util_foreach_probe_v1 (db, append_probe, &clos);
+
+      gvdb_table_free (db);
+
+      i += 1;
+    }
+
+  g_autoptr(JsonNode) json_res = NULL;
+  g_autoptr(GString) buf = NULL;
+
+  if (g_strcmp0 (opt_format, "json") == 0)
+    {
+      json_res = json_node_new (JSON_NODE_ARRAY);
+
+      JsonArray *arr = json_array_new ();
+      json_node_take_array (json_res, arr);
+    }
+  else if (g_strcmp0 (opt_format, "plain") == 0)
+    {
+      buf = g_string_new (NULL);
+    }
+
+  GHashTableIter iter;
+  gpointer value;
+
+  g_hash_table_iter_init (&iter, probes);
+  while (g_hash_table_iter_next (&iter, NULL, &value))
+    {
+      ProbeData *p = value;
+      JsonArray *res_array = NULL;
+      JsonArray *avg_array = NULL;
+      JsonObject *obj = NULL;
+
+      if (json_res != NULL)
+        {
+          res_array = json_node_get_array (json_res);
+
+          obj = json_object_new ();
+          json_array_add_object_element (res_array, obj);
+
+          json_object_set_string_member (obj, "probeName", p->probe_name);
+
+          avg_array = json_array_sized_new (p->results->len);
+          json_object_set_array_member (obj, "averageResults", avg_array);
+        }
+      else
+        {
+          g_string_append_printf (buf,
+                                  "Probe: %s\n"
+                                  "  ┕━ • avg: ",
+                                  p->probe_name);
+        }
+
+      for (int j = 0; j < p->results->len; j++)
+        {
+          ProbeResult *r = &g_array_index (p->results, ProbeResult, j);
+
+          if (avg_array != NULL)
+            {
+              json_array_add_double_element (avg_array, r->avg);
+            }
+          else
+            {
+              g_string_append_printf (buf, "%.02f %s", scale_val (r->avg), unit_for (r->avg));
+
+              if (fabs (r->avg - p->avg) < FLT_EPSILON)
+                g_string_append (buf, "[=]");
+              else if (r->avg > p->avg)
+                g_string_append (buf, "[+]");
+              else if (r->avg < p->avg)
+                g_string_append (buf, "[-]");
+
+              if (j == p->results->len - 1)
+                g_string_append (buf, "\n");
+              else
+                g_string_append (buf, ", ");
+           }
+        }
+    }
+
+  g_autofree char *data = NULL;
+
+  if (json_res != NULL)
+    {
+      g_autoptr(JsonGenerator) gen = json_generator_new ();
+
+      json_generator_set_root (gen, json_res);
+      data = json_generator_to_data (gen, NULL);
+    }
+  else if (buf != NULL)
+    {
+      data = g_string_free (buf, FALSE);
+      buf = NULL;
+    }
+
+  write (output_fd, data, strlen (data));
+  if (output_fd != STDOUT_FILENO)
+    {
+      close (output_fd);
+
+      if (rename (output_tmpfile, opt_output) != 0)
```

Comment:
What's the reason for doing this via `rename()` instead of writing to the output file directly? It will fail if `/tmp` and the output file are on different filesystems.

Reply:
The idea is to always use an atomic write+rename; it's basically `g_file_set_contents()`. I should definitely handle errors a bit better, though.

Reply:
:+1: Are you planning to add the error handling as part of this PR?

Reply:
I did.